packages/api/src/clickhouse/index.ts

```diff
@@ -904,6 +912,588 @@ export const getMetricsChart = async ({
+      SELECT ?
+      FROM ??
+      WHERE ? AND (?) ? ?
+      GROUP BY ?
+      ORDER BY ts_bucket ASC
+      ${
+        granularity != null
+          ? `WITH FILL
+        FROM toUnixTimestamp(toStartOfInterval(toDateTime(?), INTERVAL ?))
+        TO toUnixTimestamp(toStartOfInterval(toDateTime(?), INTERVAL ?))
+        STEP ?`
+          : ''
+      }${
+      sortOrder === 'asc' || sortOrder === 'desc' ? `, data ${sortOrder}` : ''
+    }
+    `,
+    [
+      SqlString.raw(selectClause),
+      tableName,
+      buildTeamLogStreamWhereCondition(tableVersion, teamId),
+      SqlString.raw(whereClause),
+      SqlString.raw(
+        !isCountFn && field != null
+          ? ` AND (${await serializer.isNotNull(field, false)})`
+          : '',
+      ),
+      SqlString.raw(
+        hasGroupBy
+          ? ` AND (${await serializer.isNotNull(groupBy, false)})`
+          : '',
+      ),
+      SqlString.raw(groupByClause),
+      ...(granularity != null
+        ? [
+            startTime / 1000,
+            granularity,
+            endTime / 1000,
+            granularity,
+            ms(granularity) / 1000,
+          ]
+        : []),
+    ],
+  );
+
+  return {
+    query,
+    hasGroupBy,
+  };
+};
+
+export const queryMultiSeriesChart = async ({
+  maxNumGroups,
+  tableVersion,
+  teamId,
+  seriesReturnType = 'column',
+  queries,
+}: {
+  maxNumGroups: number;
+  tableVersion: number | undefined;
+  teamId: string;
+  seriesReturnType?: 'ratio' | 'column';
+  queries: { query: string; hasGroupBy: boolean }[];
+}) => {
+  // For now only supports same-table series with the same groupBy
+
+  const seriesCTEs = SqlString.raw(
+    'WITH ' + queries.map((q, i) => `series_${i} AS (${q.query})`).join(',\n'),
+  );
+
+  // Only join on group bys if all queries have group bys
+  // TODO: This will not work for an array of group by fields
+  const allQueiesHaveGroupBy = queries.every(q => q.hasGroupBy);
+
+  let leftJoin = '';
+  // Join every series after the first one
+  for (let i = 1; i < queries.length; i++) {
+    leftJoin += `LEFT JOIN series_${i} AS series_${i} ON series_${i}.ts_bucket=series_0.ts_bucket${
+      allQueiesHaveGroupBy ? ` AND series_${i}.group = series_0.group` : ''
+    }\n`;
+  }
+
+  const select =
+    seriesReturnType === 'column'
+      ? queries
+          .map((_, i) => {
+            return `series_${i}.data as "series_${i}.data"`;
+          })
+          .join(',\n')
+      : 'series_0.data / series_1.data as "series_0.data"';
+
+  // Return each series data as a separate column
+  const query = SqlString.format(
+    `? 
+      ,raw_groups AS (
+        SELECT 
+          ?,
+          series_0.ts_bucket as ts_bucket, 
+          series_0.group as group
+        FROM series_0 AS series_0
+        ?
+      ), groups AS (
+        SELECT *, MAX(${
+          seriesReturnType === 'column'
+            ? `greatest(${queries
+                .map((_, i) => `series_${i}.data`)
+                .join(', ')})`
```

Comment:
style suggestion. any chance we can move this to the format 2nd arg ? (to improve readability)